env/common_properties.asciidoc

```diff
@@ -393,3 +393,120 @@ additional types in an entry point's parameter list.
 An *OpVariable* in a SPIR-V module with the *BuiltIn* decoration represents
 a built-in variable.
 All built-in variables must be in the *Input* storage class.
+
+The following table describes the required SPIR-V type for built-in variables.
+In this table, `size_t` is used as a generic type to represent:
+
+  * *OpTypeInt* with _Width_ equal to 32 if the _Addressing Model_ declared in *OpMemoryModel* is *Physical32*.
+  * *OpTypeInt* with _Width_ equal to 64 if the _Addressing Model_ declared in *OpMemoryModel* is *Physical64*.
+
+The mapping from an OpenCL C built-in function is informational and non-normative.
+
+[cols="2,2,3",options="header"]
+|====
+|*OpenCL C Function*
+|*SPIR-V BuiltIn*
+|*Required SPIR-V Type*
+
+| `get_num_groups`
+ | *NumWorkgroups*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_local_size`
+ | *WorkgroupSize*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_group_id`
+ | *WorkgroupId*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_num_groups`
+ | *NumWorkgroups*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_local_size`
+ | *WorkgroupSize*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_group_id`
+ | *WorkgroupId*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_local_id`
+ | *LocalInvocationId*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_global_id`
+ | *GlobalInvocationId*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_local_linear_id`
+ | *LocalInvocationIndex*
+  | `size_t`
+
+| `get_work_dim`
+ | *WorkDim*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_global_size`
+ | *GlobalSize*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_enqueued_local_size`
+ | *EnqueuedWorkgroupSize*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_global_offset`
+ | *GlobalOffset*
+  | *OpTypeVector* of 3 components of `size_t`
+
+| `get_global_linear_id`
+ | *GlobalLinearId*
+  | `size_t`
+
+| `get_sub_group_size`
+ | *SubgroupSize*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_max_sub_group_size`
+ | *SubgroupMaxSize*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_num_sub_groups`
+ | *NumSubgroups*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_enqueued_num_sub_groups`
+ | *NumEnqueuedSubgroups*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_sub_group_id`
+ | *SubgroupId*
+  | *OpTypeInt* with _Width_ equal to 32
+
+| `get_sub_group_local_id`
+ | *SubgroupLocalInvocationId*
+  | *OpTypeInt* with _Width_ equal to 32
+
+// TODO: Ensure this is documented as part of the extension!
```

Comment:
It is, though as a paragraph rather than a table.

Reply:
Hmm, I wrote the above a month ago but I can't find what made me say it.  The extension specification describes the non-normative mapping between OpenCL C and SPIR-V, but not the type of the SPIR-V builtin variables.

So this could be uncommented now?

Reply:
They're listed at the very bottom of [this section](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_cl_khr_subgroup_ballot) in text.

We need to decide if this table should only contain "core" BuiltIns or if it should contain extension BuiltIns also.  If it should only contain core BuiltIns then I think it should be merged as-is (I could delete the comments).  If it should contain extension BuiltIns also, are any others missing?

Reply:
> They're listed at the very bottom of this section in text.

Ah, thanks.

> We need to decide if this table should only contain "core" BuiltIns or if it should contain extension BuiltIns also. If it should only contain core BuiltIns then I think it should be merged as-is (I could delete the comments). If it should contain extension BuiltIns also, are any others missing?

I think it would be okay to include them in this table, other extension-only information, such as Float16 ULPs, exist alongside non-extension information.  But if that meant deleting the mentioned paragraph in the `cl_khr_subgroup_ballot` section to avoid duplication then probably best to not add to the table, but keep all the information together in one section.

Reply:
Merge this as is?

> If it should contain extension BuiltIns also, are any others missing?

From reading the [builtin variables table](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_builtin_a_builtin), no, I do not believe any more apply to OpenCL.